ALB support

atrope opened the pull request:
Adding ALB Support for those who wish to avoid APIGW pricing and/or any other reason to use ALB instead of APIGW(Automatic Health check makes ourlambdas always hot  :)). 

PS: This PR also needs to consider the https://github.com/hoaproject/Fastcgi/pull/28 that enables Responder class to process multi headers(that we must enable in our target group on ALBs).

deleugpn commented:
Has SAM / CloudFormation included support for Lambda in the Target Group of a Load Balancer Listener or are you defining it just on AWS Web Console?

atrope commented:
@deleugpn AFAIK they did not. I Did it on Console

atrope commented:
@mnapoli I Updated from master and added the new code for Hoa's multi header.
I Dont know why the codes are not passing if you can take a look will be great.

All failed test cases are the same for what i checked.

atrope commented:
@mnapoli Updated the PR to use Hollodotme.
The getHeaders function is necessary while Hollodotme dont fully support multiheaders, but we have no reason to wait for that because we can parse ourselves.

Please take a look :+1:

atrope commented on src/Http/LambdaResponse.php:
```diff
@@ -71,4 +72,19 @@ public function toApiGatewayFormat(): array
+    {
+        // This is the format required by the ALB lambda integration https://aws.amazon.com/pt/blogs/networking-and-content-delivery/lambda-functions-as-targets-for-application-load-balancers/
+        $statusDescription = "$this->statusCode ";
+        $statusDescription .= $this->http_codes[$this->statusCode] ?? 'Unknown';
```
I Agree.
While i am sure it was a must back then, i just tested and without even sending the statusDescription property it worked without problems. so i am removing it.
AWS must have updated its ALB so it would add it by default.

atrope commented on src/Runtime/LambdaRuntime.php:
```diff
@@ -255,10 +259,12 @@ private function postJson(string $url, $data): void
+                'Content-Type: application/json',
+                'Content-Length: ' . strlen($jsonData),
+            ]);
+        }
```
Removed the If. This was an workaround when using the other layers that it is not needed now.

atrope commented on src/Runtime/PhpFpm.php:
```diff
@@ -99,7 +99,33 @@ public function ensureStillRunning(): void
+            $responseHeaders = $response->getHeaders();
+        }
+        return array_change_key_case($responseHeaders, CASE_LOWER);
+    }
```
MultiHeader is an absolute must for ALB.
But not only for ALB.
I Refactored every to not take into account ALB/ApiGW

Now it looks for the multivalueheader and does everything it should do.

atrope commented:
@mnapoli i changed the ToApiGateway method as it does not makes sense anymore (it serves also for ALB) please take a look, it does makes more sense now, and the PR should be "Adding Multiheaders Support (ALB)". I've tested in a working environment and all was good..

hmazter commented:
Maybe a little off topic here. But I saw today that there is an option for the ALB Target group to switch between using multi or single values for headers.
Don't know if its new or not.

* https://docs.aws.amazon.com/elasticloadbalancing/latest/application/lambda-functions.html#enable-multi-value-headers
* https://www.terraform.io/docs/providers/aws/r/lb_target_group.html#lambda_multi_value_headers_enabled
* https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-elasticloadbalancingv2-targetgroup-targetgroupattribute.html

atrope commented:
@hmazter In this last refactor that i did, it takes into account in every request if it is Multi Valued or not it does not check if the request comes  from APIGW or ALB

mnapoli commented:
Thanks @atrope, in order to simplify the review and move forward I've opened #338 that includes your PR and adds the tests.

mnapoli commented:
This is not obvious by the message about but for anyone reading this in the future: this has been merged via #338
